File: app/lib/domain/entities/ride.dart
import 'package:equatable/equatable.dart';
import 'package:ride_journal/domain/entities/route_point.dart';

enum RideVisibility { private_, followers, public_ }

class Ride extends Equatable {
  final String id;
  final String name;
  final double distanceMeters;
  final Duration duration;
  final double avgSpeedKmh;
  final double maxSpeedKmh;
  final double elevationGainMeters;
  final DateTime startTime;
  final DateTime? endTime;
  final List<RoutePoint> routePoints;
  final RideVisibility visibility;
  final DateTime updatedAt;
  final DateTime? syncedAt;

  const Ride({
    required this.id,
    required this.name,
    required this.distanceMeters,
    required this.duration,
    required this.avgSpeedKmh,
    required this.maxSpeedKmh,
    required this.elevationGainMeters,
    required this.startTime,
    this.endTime,
    required this.routePoints,
    this.visibility = RideVisibility.private_,
    required this.updatedAt,
    this.syncedAt,
  });

  bool get isActive => endTime == null;

  Ride copyWith({
    String? id,
    String? name,
    double? distanceMeters,
    Duration? duration,
    double? avgSpeedKmh,
    double? maxSpeedKmh,
    double? elevationGainMeters,
    DateTime? startTime,
    DateTime? endTime,
    List<RoutePoint>? routePoints,
    RideVisibility? visibility,
    DateTime? updatedAt,
    DateTime? syncedAt,
  }) {
    return Ride(
      id: id ?? this.id,
      name: name ?? this.name,
      distanceMeters: distanceMeters ?? this.distanceMeters,
      duration: duration ?? this.duration,
      avgSpeedKmh: avgSpeedKmh ?? this.avgSpeedKmh,
      maxSpeedKmh: maxSpeedKmh ?? this.maxSpeedKmh,
      elevationGainMeters: elevationGainMeters ?? this.elevationGainMeters,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      routePoints: routePoints ?? this.routePoints,
      visibility: visibility ?? this.visibility,
      updatedAt: updatedAt ?? this.updatedAt,
      syncedAt: syncedAt ?? this.syncedAt,
    );
  }

  @override
  List<Object?> get props => [
    id,
    name,
    distanceMeters,
    duration,
    avgSpeedKmh,
    maxSpeedKmh,
    elevationGainMeters,
    startTime,
    endTime,
    routePoints,
    visibility,
    updatedAt,
    syncedAt,
  ];
}

File: backend-api/src/RideJournal.Domain/Entities/Ride.cs
using RideJournal.Domain.Enums;

namespace RideJournal.Domain.Entities;

public class Ride
{
    public string Id { get; set; } = string.Empty;
    public string UserId { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public double DistanceMeters { get; set; }
    public long DurationMs { get; set; }
    public double AvgSpeedKmh { get; set; }
    public double MaxSpeedKmh { get; set; }
    public double ElevationGainMeters { get; set; }
    public DateTime StartTime { get; set; }
    public DateTime? EndTime { get; set; }
    public List<RoutePoint> RoutePoints { get; set; } = new();
    public RideVisibility Visibility { get; set; } = RideVisibility.Private;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// The client's logical version of this ride, used only for conflict
    /// resolution (last writer by client clock wins).
    /// </summary>
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Server-clock write timestamp, stamped by the repository on every write.
    /// This is the sync high-water mark: <see cref="UpdatedAt"/> comes from the
    /// pushing device, so a device whose clock runs behind the server would
    /// silently never propagate to other devices if the cursor used it.
    /// </summary>
    public DateTime ServerUpdatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Soft-delete tombstone. Deletes must survive long enough to propagate to
    /// other devices, so rows are never removed outright.
    /// </summary>
    public DateTime? DeletedAt { get; set; }
}
